minimint/src/net/peers.rs

```diff
@@ -0,0 +1,457 @@
+                    debug!(?peer, "Trying to reconnect");
+                    assert!(connection.is_none());
+                    let (connected_peer, mut conn) = match connect.connect_framed(cfg.addr.clone()).await {
+                        Ok(peer_conn) => peer_conn,
+                        Err(e) => {
+                            warn!(?peer, addr = ?cfg.addr, err = %e, "Connecting to peer failed");
+                            err = Some(e);
+                            continue;
+                        }
+                    };
+
+                    if connected_peer != peer {
+                        error!(expected = ?peer, got = ?connected_peer, "Peer identified itself incorrectly");
+                        err = Some(anyhow::anyhow!("Peer identified itself incorrectly"));
+                        continue;
+                    }
+
+                    for msg in resend_queue.iter().cloned() {
+                        if let Err(e) = conn.send(PeerMessage {
+                            msg,
+                            ack: last_received
+                        }).await {
+                            err = Some(e);
+                            continue;
+                        }
+                    }
+
+                    if connection.replace(conn).is_some() {
+                        warn!(?peer, "Replaced old connection");
+                    }
+                    failed_reconnect_counter = 0;
+                    reconnect = None;
+                }
+            };
+        }
+    }
+}
+
+async fn next_if_some<S>(stream: &mut Option<S>) -> Option<S::Item>
+where
+    S: Stream + Unpin,
+{
+    match stream.as_mut() {
+        Some(stream) => stream.next().await,
+        None => std::future::pending().await,
+    }
+}
+
+async fn sleep_if_some(instant: Option<Instant>) {
+    match instant {
+        Some(deadline) => tokio::time::sleep_until(deadline).await,
+        None => std::future::pending().await,
+    }
+}
+
+fn reconnect_time(failed_reconnect_counter: u64) -> Instant {
+    let scaling_factor = failed_reconnect_counter as f64;
+    let delay: f64 = thread_rng().gen_range(1.0 * scaling_factor, 4.0 * scaling_factor);
+    debug!(delay, "Scheduling reopening of connection");
+    Instant::now() + Duration::from_secs_f64(delay)
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::net::connect::mock::MockNetwork;
+    use crate::net::connect::Connector;
+    use crate::net::peers::{
+        ConnectionConfig, NetworkConfig, PeerConnections, ReconnectPeerConnections,
+    };
+    use futures::Future;
+    use hbbft::Target;
+    use minimint_api::PeerId;
+    use std::collections::HashMap;
+    use std::time::Duration;
+    use tracing_subscriber::EnvFilter;
+
+    async fn timeout<F, T>(f: F) -> Option<T>
+    where
+        F: Future<Output = T>,
+    {
+        tokio::time::timeout(Duration::from_secs(100), f).await.ok()
+    }
+
+    #[tokio::test]
+    async fn test_connect() {
```

Comment:
a separate disconnect test would be cool eventually, but that needs a more versatile `MockNetwork` impl that supports killing connections.